docker/provider.go

```diff
@@ -42,6 +48,45 @@ func Provider() terraform.ResourceProvider {
 				DefaultFunc: schema.EnvDefaultFunc("DOCKER_CERT_PATH", ""),
 				Description: "Path to directory with Docker TLS config",
 			},
+
+			"registry_auth": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"address": &schema.Schema{
+							Type:        schema.TypeString,
+							Required:    true,
+							Description: "Address of the registry",
+						},
+
+						"username": &schema.Schema{
+							Type:          schema.TypeString,
+							Optional:      true,
+							ConflictsWith: []string{"registry_auth.config_file"},
+							DefaultFunc:   schema.EnvDefaultFunc("DOCKER_REGISTRY_USER", ""),
+							Description:   "Username for the registry",
+						},
+
+						"password": &schema.Schema{
+							Type:          schema.TypeString,
+							Optional:      true,
+							ConflictsWith: []string{"registry_auth.config_file"},
+							DefaultFunc:   schema.EnvDefaultFunc("DOCKER_REGISTRY_PASS", ""),
+							Description:   "Password for the registry",
+						},
+
+						"config_file": &schema.Schema{
+							Type:          schema.TypeString,
+							Optional:      true,
+							ConflictsWith: []string{"registry_auth.username", "registry_auth.password"},
+							DefaultFunc:   schema.EnvDefaultFunc("DOCKER_CONFIG", "~/.docker/config.json"),
+							Description:   "Path to docker json file for registry auth",
+						},
+					},
+				},
+				Set: providerRegistryAuthHash,
```

Comment:
Unless you _really_ need to define this `Set` id method yourself, you should not. Omitting `Set` will use the default for this `Set`'s `Elem` type:

https://github.com/terraform-providers/terraform-provider-aws/blob/a301ad908ed20a528c338cc59a8e5122d753910a/vendor/github.com/hashicorp/terraform/helper/schema/schema.go#L273-L294

Reply:
Alright, so do you have a good case when it is _really_ necessary to define this method? Maybe this one in the `aws-provider` 🤔  Because it has a nested `Map` type?

https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/data_source_aws_ami.go#L129

Reply:
That example probably pre-dates the built-in `HashSchema` method, sadly 😦 

I do not have an example handy, so if you could, just use the default :D